File: frontend/src/kpis/canales.ts
/*
 * Adquisición — espejo de backend/api/kpis/canales.py.
 *
 * Volumen al lado de la tasa, siempre: un canal de 300 leads al 25% y uno de
 * 15.000 al 4,5% no se comparan por la tasa sola.
 *
 * Cuatro cortes de la misma pregunta: el agregado (`canales`), el mismo corte
 * mes a mes (`canalesMes`), el canal abierto por su fuente concreta
 * (`origenes`) y solo el canal pagado (`campanas`, que ya no alimenta ninguna
 * pantalla: `origenes` lo contiene, con el nombre de la campaña en vez del id).
 */
import type { CampanaRow, CanalMesRow, CanalRow, OrigenRow, Snapshot } from '../types'
import { pct, promedio, red } from './base'

export function canales(S: Snapshot, sel: number[]): CanalRow[] {
  const c = S.col
  const g = new Map<number, { leads: number; ganados: number; td: number; dias: number[]; msg: number }>()
  for (const i of sel) {
    const k = c.ca[i]
    if (k < 0) continue
    let x = g.get(k)
    if (!x) g.set(k, (x = { leads: 0, ganados: 0, td: 0, dias: [], msg: 0 }))
    x.leads++; x.ganados += c.g[i]; x.td += c.td[i]; x.msg += c.nmsg[i]
    if (c.dias[i] >= 0) x.dias.push(c.dias[i])
  }
  return [...g].map(([k, x]) => ({
    canal: S.dic.canal[k], leads: x.leads, ganados: x.ganados,
    tasa_conversion: pct(x.ganados, x.leads),
    tasa_test_drive: pct(x.td, x.leads),
    dias_promedio: red(promedio(x.dias), 1),
    mensajes_promedio: red(x.msg / x.leads, 1),
  })).sort((a, b) => b.leads - a.leads)
}

export function campanas(S: Snapshot, sel: number[]): CampanaRow[] {
  const c = S.col
  const g = new Map<number, { leads: number; ganados: number; td: number; pl: number; ob: number }>()
  for (const i of sel) {
    const k = c.cp[i]
    if (k < 0) continue
    let x = g.get(k)
    if (!x) g.set(k, (x = { leads: 0, ganados: 0, td: 0, pl: c.pl[i], ob: c.ob[i] }))
    x.leads++; x.ganados += c.g[i]; x.td += c.td[i]
  }
  return [...g].map(([k, x]) => ({
    campana: S.dic.campana[k], platform: S.dic.platform[x.pl], objetivo: S.dic.objetivo[x.ob],
    leads: x.leads, ganados: x.ganados,
    tasa_conversion: pct(x.ganados, x.leads),
    tasa_test_drive: pct(x.td, x.leads),
  })).sort((a, b) => b.leads - a.leads)
}

/**
 * La serie mensual de cada canal. Espejo de `canales_mes` en SQL.
 *
 * `medible` se calcula por MES sobre TODA la población filtrada —antes de
 * separar por canal— igual que en la query, donde el CTE `edad` agrupa `pob`
 * solo por mes. Calcularla dentro de cada canal daría cuatro cortes distintos
 * para el mismo mes.
 */
export function canalesMes(S: Snapshot, sel: number[]): CanalMesRow[] {
  const c = S.col
  // Contra el corte del DATASET, no el del filtro: mirar sep–dic 2025 no vuelve
  // incomparables a esas cohortes, ya pasó el tiempo real. Igual que en `cohortes`.
  const corte = S.contexto.dia_corte
  const ultimo = new Map<number, number>()
  const g = new Map<string, {
    me: number; ca: number; leads: number; ganados: number; ventana: number
  }>()
  for (const i of sel) {
    const me = c.me[i]
    if (c.dia[i] > (ultimo.get(me) ?? -1)) ultimo.set(me, c.dia[i])
    const ca = c.ca[i]
    if (ca < 0) continue
    const k = `${me}|${ca}`
    let x = g.get(k)
    if (!x) g.set(k, (x = { me, ca, leads: 0, ganados: 0, ventana: 0 }))
    x.leads++; x.ganados += c.g[i]; x.ventana += c.gv[i]
  }
  return [...g.values()].map((x) => ({
    mes: S.dic.mes[x.me], canal: S.dic.canal[x.ca],
    leads: x.leads, ganados: x.ganados,
    tasa_conversion: pct(x.ganados, x.leads),
    ganados_ventana: x.ventana,
    tasa_ventana: pct(x.ventana, x.leads),
    medible: corte - (ultimo.get(x.me) ?? corte) >= S.contexto.ventana_dias,
  })).sort((a, b) => (a.mes === b.mes ? (a.canal < b.canal ? -1 : 1) : a.mes < b.mes ? -1 : 1))
}

/** El canal abierto por la fuente concreta del lead. Espejo de `origenes`. */
export function origenes(S: Snapshot, sel: number[]): OrigenRow[] {
  const c = S.col
  const g = new Map<number, {
    ca: number; cp: number; leads: number; ganados: number; td: number
    dias: number[]; msg: number
  }>()
  for (const i of sel) {
    const k = c.og[i]
    if (k < 0) continue
    let x = g.get(k)
    if (!x) g.set(k, (x = { ca: c.ca[i], cp: c.cp[i], leads: 0, ganados: 0, td: 0, dias: [], msg: 0 }))
    x.leads++; x.ganados += c.g[i]; x.td += c.td[i]; x.msg += c.nmsg[i]
    if (c.dias[i] >= 0) x.dias.push(c.dias[i])
  }
  return [...g].map(([k, x]) => ({
    canal: S.dic.canal[x.ca],
    origen: S.dic.origen[k],
    // el COALESCE(campana, origen_detalle) de la query
    rotulo: x.cp >= 0 ? S.dic.campana[x.cp] : S.dic.origen[k],
    leads: x.leads, ganados: x.ganados,
    tasa_conversion: pct(x.ganados, x.leads),
    tasa_test_drive: pct(x.td, x.leads),
    dias_promedio: red(promedio(x.dias), 1),
    mensajes_promedio: red(x.msg / x.leads, 1),
  })).sort((a, b) => b.leads - a.leads)
}

File: backend/api/kpis/canales.py
"""
Adquisición: de dónde vienen los leads y cuáles valen la pena.

La pregunta #1 de la gerencia, y la que más fácil se contesta mal: la tasa sola
sin el volumen al lado hace que un canal de 300 leads parezca la estrategia.

Cuatro cortes de la misma pregunta, de menos a más fino:

    canales      el agregado del período, cuatro filas
    canales_mes  el mismo corte MES A MES — sin esto no se ve que el canal
                 pagado cayó de 7,2% a 1,5% mientras subía el volumen
    origenes     el canal abierto por su fuente concreta, que es donde vive la
                 diferencia de 2x adentro de un mismo canal
    campanas     solo el canal pagado; lo subsume `origenes` y queda como
                 endpoint suelto
"""
from __future__ import annotations

from filtros import Filtros

from .base import filas


def canales(cur, f: Filtros) -> list[dict]:
    """Qué canal vale la pena.

    Va con volumen al lado de la tasa a propósito: un canal con 25% de
    conversión y 300 leads no es comparable con uno de 4,5% y 15.000.
    """
    w, p = f.where()
    return filas(cur, f"""
        SELECT l.canal,
               COUNT(*)                                        AS leads,
               COUNT(*) FILTER (WHERE l.es_ganado)              AS ganados,
               ROUND(100.0 * AVG(l.es_ganado::int), 2)          AS tasa_conversion,
               ROUND(100.0 * AVG(l.test_drive::int), 2)         AS tasa_test_drive,
               ROUND(AVG(l.dias_en_embudo), 1)                  AS dias_promedio,
               ROUND(AVG(l.n_mensajes), 1)                      AS mensajes_promedio
          FROM analytics.fct_leads l WHERE {w}
         GROUP BY 1 ORDER BY leads DESC""", p)


def campanas(cur, f: Filtros) -> list[dict]:
    """Desagrega el canal pagado. 'Los ads convierten mal' no es accionable;
    'la campaña lookalike trae el 50% del volumen pagado al 3,8%' sí lo es."""
    w, p = f.where()
    return filas(cur, f"""
        SELECT l.platform, l.campana, l.objetivo,
               COUNT(*)                                        AS leads,
               COUNT(*) FILTER (WHERE l.es_ganado)              AS ganados,
               ROUND(100.0 * AVG(l.es_ganado::int), 2)          AS tasa_conversion,
               ROUND(100.0 * AVG(l.test_drive::int), 2)         AS tasa_test_drive
          FROM analytics.fct_leads l
         WHERE {w} AND l.campana IS NOT NULL
         GROUP BY 1, 2, 3 ORDER BY leads DESC""", p)


def canales_mes(cur, f: Filtros, ventana_dias: float) -> list[dict]:
    """La serie mensual de cada canal: volumen y conversión, mes a mes.

    Es la pregunta de presupuesto que el agregado no puede contestar. `canales`
    dice que el canal pagado convierte al 4,5%; esto dice que venía del 7,2% y
    terminó en 1,5% mientras subía el volumen. Un promedio de doce meses esconde
    exactamente el movimiento sobre el que se decide.

    La ventana de observación es la MISMA de `cohortes` y por la misma razón: el
    resultado final de un mes reciente está incompleto, así que las cuatro líneas
    terminan en un derrumbe que solo mide la edad de las cohortes. Acá se cuenta
    la venta que ocurrió dentro de los primeros `ventana_dias` del lead, que es
    la única cifra que se puede poner al lado de la de otro mes.

    `medible` se decide por MES y no por mes×canal a propósito: así las líneas de
    todos los canales se cortan en el mismo punto. Con un corte por canal, el
    límite de cada línea dependería de a qué hora entró su último lead — mismo
    dato, cuatro cortes distintos y ninguno explicable.
    """
    w, p = f.where()
    return filas(cur, f"""
        WITH corte AS (
            SELECT MAX(creado_en) AS hasta FROM analytics.fct_leads WHERE dataset_id = %s
        ),
        pob AS (
            SELECT l.mes, l.canal, l.es_ganado, l.creado_en,
                   -- COALESCE: sin él un NULL entraría al AVG como fila
                   -- descartada y encogería el denominador en silencio
                   COALESCE(l.es_ganado AND l.dias_a_ganado <= %s, false) AS en_ventana
              FROM analytics.fct_leads l WHERE {w}
        ),
        edad AS (
            SELECT mes,
                   (EXTRACT(EPOCH FROM (SELECT hasta FROM corte) - MAX(creado_en)) / 86400.0)
                       >= %s                                   AS medible
              FROM pob GROUP BY 1
        )
        SELECT p.mes, p.canal,
               COUNT(*)                                        AS leads,
               COUNT(*) FILTER (WHERE p.es_ganado)              AS ganados,
               ROUND(100.0 * AVG(p.es_ganado::int), 2)          AS tasa_conversion,
               COUNT(*) FILTER (WHERE p.en_ventana)             AS ganados_ventana,
               ROUND(100.0 * AVG(p.en_ventana::int), 2)         AS tasa_ventana,
               e.medible
          FROM pob p JOIN edad e USING (mes)
         GROUP BY 1, 2, 8 ORDER BY 1, 2""",
                 [f.dataset_id, ventana_dias] + p + [ventana_dias])


def origenes(cur, f: Filtros) -> list[dict]:
    """El canal, abierto por el origen concreto del lead.

    `canal` es la categoría contable —pagado, orgánico, formulario, outbound— y
    adentro de cada una hay fuentes que no se parecen en nada: en orgánico, el
    link de WhatsApp convierte al 34,1% y el botón del sitio al 17,2%, y el
    promedio del canal (24,7%) no describe a ninguno de los dos. Repartir
    presupuesto sobre ese promedio es repartirlo entre una fuente que rinde el
    doble que la otra como si rindieran igual.

    `rotulo` es el nombre de la campaña cuando el origen es un aviso pagado —el
    id numérico de Meta no le dice nada a nadie— y el slug del origen en el
    resto. Con eso esta query subsume a `campanas`: los tres orígenes del canal
    pagado SON las tres campañas, y dibujarlas en una tabla aparte contaba lo
    mismo dos veces. `campanas` sigue existiendo como endpoint suelto.
    """
    w, p = f.where()
    return filas(cur, f"""
        SELECT l.canal,
               l.origen_detalle                                AS origen,
               COALESCE(l.campana, l.origen_detalle)            AS rotulo,
               COUNT(*)                                        AS leads,
               COUNT(*) FILTER (WHERE l.es_ganado)              AS ganados,
               ROUND(100.0 * AVG(l.es_ganado::int), 2)          AS tasa_conversion,
               ROUND(100.0 * AVG(l.test_drive::int), 2)         AS tasa_test_drive,
               ROUND(AVG(l.dias_en_embudo), 1)                  AS dias_promedio,
               ROUND(AVG(l.n_mensajes), 1)                      AS mensajes_promedio
          FROM analytics.fct_leads l
         WHERE {w} AND l.origen_detalle IS NOT NULL
         GROUP BY 1, 2, 3 ORDER BY leads DESC""", p)
